packages/@glimmer/ssr/src/application.ts

```diff
@@ -0,0 +1,89 @@
+import { BaseApplication, Loader, Renderer } from '@glimmer/application';
+import { Resolver, Dict } from '@glimmer/di';
+import Environment from './environment';
+import { Opaque } from '@glimmer/util';
+import StringBuilder from './string-builder';
+import { Document, HTMLSerializer, voidMap } from 'simple-dom';
+import { PathReference, ConstReference } from '@glimmer/reference';
+import { PassThrough } from 'stream';
+import { ComponentManager } from '@glimmer/component';
+
+export interface SSRApplicationOptions {
+  rootName: string;
+  resolver: Resolver;
+  loader: Loader;
+  renderer: Renderer;
+}
+
+/**
+ * Converts a POJO into a dictionary of references that can be passed as an argument to render a component.
+ */
+function convertOpaqueToReferenceDict(data: Opaque): Dict<PathReference<Opaque>> {
+  if (!data) {
+    return {};
+  };
+
+  return Object.keys(data).reduce((acc, key) => {
+    acc[key] = new ConstReference(data[key]);
+    return acc;
+  }, {});
+}
+
+export default class Application extends BaseApplication {
+  protected serializer: HTMLSerializer;
+
+  constructor({rootName, resolver, loader, renderer}: SSRApplicationOptions) {
+    super({
+      rootName,
+      resolver,
+      loader,
+      renderer,
+      environment: Environment
+    });
+
+    this.serializer = new HTMLSerializer(voidMap);
+    this.registerInitializer({
+      initialize(registry) {
+        registry.register(`component-manager:/${rootName}/component-managers/main`, ComponentManager);
+      }
+    });
+
+    // Setup registry and DI
+    this.initialize();
```

Comment:
Maybe I missed it, but don't we need to create/destroy the container per request? Having a shared container makes it super easy to leak PII between requests.